File: benchmarks/linalg-svd/golden/verify.ts
// =============================================================================
// linalg-svd verifier — invariant-based, language-neutral.
// =============================================================================
//
// stdin:
//   { input:     { A: number[][], mode?: "reduced" | "complete" },
//     candidate: { U, S, Vt, mode, reconstruction_error,
//                  orthogonality_error_U, orthogonality_error_Vt,
//                  condition_number, rank_estimate, method, warnings },
//     id?:       string }
//
// stdout:
//   { pass: bool, reason: str, checks: { <name>: { pass, detail } } }
//
// Runs 8 independent checks per case (see verifier_protocol.md):
//   1. shape                       — U, S, Vt dimensions match (m, n, mode)
//   2. finite_entries              — no NaN, no ±Inf in factors or scalars
//   3. S_nonneg_descending         — singular values ≥ 0 and non-increasing
//   4. U_orthonormal               — ||UᵀU − I|| within tol_orth
//   5. Vt_orthonormal              — ||Vt·Vtᵀ − I|| within tol_orth
//   6. factorisation_residual      — ||U·diag(S)·Vt − A||_F within tol_recon
//   7. self_reported_residual      — candidate's recon claim is honest
//   8. self_reported_orthogonality — candidate's orth claims (U + Vt) honest
//
// TS port of the Python verifier from scientist-workbench/bench/linalg-svd.
// Tolerances match byte-for-byte (verify.py values preserved per ADR-0028 §4
// — do not tighten or loosen during migration).
//
// The matrix operations (transpose, matmul, Frobenius norm) are inlined
// because the tournament-protocol contract is "single executable, JSON in,
// JSON out" — no external linalg dependency.

const EPS                 = 2.220446049250313e-16;
const SAFETY              = 100.0;
const SELF_REPORT_REL_TOL = 1e-6;

// ─── small linalg helpers ────────────────────────────────────────────────────

function transpose(A: number[][]): number[][] {
  const m = A.length;
  const n = m === 0 ? 0 : (A[0]?.length ?? 0);
  const out: number[][] = Array.from({ length: n }, () => new Array<number>(m).fill(0));
  for (let i = 0; i < m; i++) {
    const row = A[i]!;
    for (let j = 0; j < n; j++) out[j]![i] = row[j]!;
  }
  return out;
}

function matmul(A: number[][], B: number[][]): number[][] {
  const m = A.length;
  const k = m === 0 ? 0 : (A[0]?.length ?? 0);
  const n = B.length === 0 ? 0 : (B[0]?.length ?? 0);
  const out: number[][] = Array.from({ length: m }, () => new Array<number>(n).fill(0));
  for (let i = 0; i < m; i++) {
    const arow = A[i]!;
    const orow = out[i]!;
    for (let p = 0; p < k; p++) {
      const aip = arow[p]!;
      if (aip === 0) continue;
      const brow = B[p]!;
      for (let j = 0; j < n; j++) orow[j]! += aip * brow[j]!;
    }
  }
  return out;
}

function frobNorm(A: number[][]): number {
  let s = 0;
  for (const row of A) for (const v of row) s += v * v;
  return Math.sqrt(s);
}

function matSub(A: number[][], B: number[][]): number[][] {
  return Array.from({ length: A.length }, (_, i) =>
    A[i]!.map((v, j) => v - B[i]![j]!),
  );
}

// Build the rectangular Sigma matrix: q_U × q_Vt, with S along the diagonal.
function buildSigma(S: number[], qU: number, qVt: number): number[][] {
  const out: number[][] = Array.from({ length: qU }, () => new Array<number>(qVt).fill(0));
  for (let i = 0; i < S.length; i++) out[i]![i] = S[i]!;
  return out;
}

// UᵀU − I_q (gram matrix minus identity, for orthonormality check).
function gramMinusEye(M: number[][]): number[][] {
  const Mt = transpose(M);
  const gram = matmul(Mt, M);
  const q = gram.length;
  for (let i = 0; i < q; i++) gram[i]![i]! -= 1;
  return gram;
}

// Vt · Vtᵀ − I_q (row-gram matrix minus identity, for Vt orthonormality).
function rowGramMinusEye(Vt: number[][]): number[][] {
  const q = Vt.length;
  const VtT = transpose(Vt);
  const gram = matmul(Vt, VtT);
  for (let i = 0; i < q; i++) gram[i]![i]! -= 1;
  return gram;
}

function shapeOf(mat: unknown): [number, number] | null {
  if (!Array.isArray(mat)) return null;
  if (mat.length === 0) return [0, 0];
  for (const row of mat) if (!Array.isArray(row)) return null;
  const ncols = (mat[0] as number[]).length;
  for (const row of mat as number[][]) {
    if (row.length !== ncols) return null;
    for (const v of row) if (typeof v !== "number") return null;
  }
  return [mat.length, ncols];
}

function vecLen(v: unknown): number | null {
  if (!Array.isArray(v)) return null;
  for (const x of v) if (typeof x !== "number") return null;
  return v.length;
}

function allFiniteMat(A: unknown): boolean {
  if (!Array.isArray(A)) return false;
  for (const row of A) {
    if (!Array.isArray(row)) return false;
    for (const v of row) if (typeof v !== "number" || !Number.isFinite(v)) return false;
  }
  return true;
}

function allFiniteVec(v: unknown): boolean {
  if (!Array.isArray(v)) return false;
  for (const x of v) if (typeof x !== "number" || !Number.isFinite(x)) return false;
  return true;
}

function fmt(x: number): string {
  if (!Number.isFinite(x)) return String(x);
  return x.toExponential(3);
}

// ─── per-check implementations ───────────────────────────────────────────────

type Check = { pass: boolean; detail: string };

function checkShape(
  A: number[][],
  cand: Record<string, unknown>,
  mode: string,
): Check {
  const m = A.length;
  const n = m === 0 ? 0 : (A[0]?.length ?? 0);
  const k = Math.min(m, n);

  // Expected shapes differ by mode.
  const expectedU: [number, number] = mode === "reduced" ? [m, k] : [m, m];
  const expectedVt: [number, number] = mode === "reduced" ? [k, n] : [n, n];
  const expectedSLen = k;

  const required = [
    "U", "S", "Vt", "mode",
    "reconstruction_error", "orthogonality_error_U",
    "orthogonality_error_Vt", "condition_number",
    "rank_estimate", "method", "warnings",
  ];
  const missing = required.filter((k) => !(k in cand));
  if (missing.length > 0) {
    return { pass: false, detail: `missing fields: [${missing.sort().join(",")}]` };
  }

  if (cand["mode"] !== mode) {
    return { pass: false, detail: `mode mismatch: candidate=${JSON.stringify(cand["mode"])}, expected=${JSON.stringify(mode)}` };
  }

  if (typeof cand["method"] !== "string") {
    return { pass: false, detail: "method must be a string" };
  }

  const warnings = cand["warnings"];
  if (!Array.isArray(warnings) || !warnings.every((s) => typeof s === "string")) {
    return { pass: false, detail: "warnings must be list[str]" };
  }

  // rank_estimate: non-negative integer (not boolean).
  const re = cand["rank_estimate"];
  if (typeof re !== "number" || typeof re === "boolean" || !Number.isInteger(re) || re < 0) {
    return { pass: false, detail: "rank_estimate must be a non-negative int" };
  }

  const sU = shapeOf(cand["U"]);
  if (!sU || sU[0] !== expectedU[0] || sU[1] !== expectedU[1]) {
    return { pass: false, detail: `U shape ${JSON.stringify(sU)}, expected ${JSON.stringify(expectedU)} for mode=${JSON.stringify(mode)}` };
  }

  const sLen = vecLen(cand["S"]);
  if (sLen !== expectedSLen) {
    return { pass: false, detail: `S length ${sLen}, expected ${expectedSLen}` };
  }

  const sVt = shapeOf(cand["Vt"]);
  if (!sVt || sVt[0] !== expectedVt[0] || sVt[1] !== expectedVt[1]) {
    return { pass: false, detail: `Vt shape ${JSON.stringify(sVt)}, expected ${JSON.stringify(expectedVt)} for mode=${JSON.stringify(mode)}` };
  }

  return { pass: true, detail: `U ${JSON.stringify(sU)}, S ${sLen}, Vt ${JSON.stringify(sVt)}, mode ${JSON.stringify(mode)}` };
}

function checkFinite(cand: Record<string, unknown>): Check {
  for (const field of ["U", "Vt"] as const) {
    if (!allFiniteMat(cand[field])) {
      return { pass: false, detail: `non-finite entry in ${field}` };
    }
  }
  if (!allFiniteVec(cand["S"])) {
    return { pass: false, detail: "non-finite entry in S" };
  }
  for (const field of ["reconstruction_error", "orthogonality_error_U",
                        "orthogonality_error_Vt", "condition_number"] as const) {
    const v = cand[field];
    if (typeof v !== "number" || !Number.isFinite(v)) {
      return { pass: false, detail: `non-finite or non-numeric ${field}` };
    }
  }
  return { pass: true, detail: "all entries finite" };
}

function checkSNonnegDescending(S: number[]): Check {
  if (S.length === 0) return { pass: true, detail: "empty S" };
  const sMax = S[0]!;
  const tolStruct = SAFETY * EPS * Math.max(sMax, 1.0);

  for (let i = 0; i < S.length; i++) {
    if (S[i]! < -tolStruct) {
      return {
        pass: false,
        detail: `S[${i}]=${fmt(S[i]!)} < -tol=${fmt(-tolStruct)} (negative singular value)`,
      };
    }
  }
  for (let i = 0; i < S.length - 1; i++) {
    if (S[i]! + tolStruct < S[i + 1]!) {
      return {
        pass: false,
        detail: `S[${i}]=${fmt(S[i]!)} < S[${i + 1}]=${fmt(S[i + 1]!)} (non-monotone, gap ${fmt(S[i + 1]! - S[i]!)} > tol=${fmt(tolStruct)})`,
      };
    }
  }
  return {
    pass: true,
    detail: `S[0]=${fmt(S[0]!)}, S[k-1]=${fmt(S[S.length - 1]!)}, monotone within tol=${fmt(tolStruct)}`,
  };
}

function checkUOrthonormal(A: number[][], U: number[][]): Check {
  const m = A.length;
  const q = U.length === 0 ? 0 : (U[0]?.length ?? 0);
  const tol = SAFETY * EPS * m * Math.sqrt(q);
  const err = frobNorm(gramMinusEye(U));
  if (err > tol) {
    return { pass: false, detail: `||UᵀU − I||_F = ${fmt(err)} > tol = ${fmt(tol)}` };
  }
  return { pass: true, detail: `||UᵀU − I||_F = ${fmt(err)} ≤ tol = ${fmt(tol)}` };
}

function checkVtOrthonormal(A: number[][], Vt: number[][]): Check {
  // Uses m (not n) for the tolerance, matching verify.py's check_Vt_orthonormal.
  const m = A.length;
  const q = Vt.length;
  const tol = SAFETY * EPS * m * Math.sqrt(q);
  const err = frobNorm(rowGramMinusEye(Vt));
  if (err > tol) {
    return { pass: false, detail: `||Vt·Vtᵀ − I||_F = ${fmt(err)} > tol = ${fmt(tol)}` };
  }
  return { pass: true, detail: `||Vt·Vtᵀ − I||_F = ${fmt(err)} ≤ tol = ${fmt(tol)}` };
}

function checkFactorisationResidual(A: number[][], U: number[][], S: number[], Vt: number[][]): Check {
  const m = A.length;
  const n = m === 0 ? 0 : (A[0]?.length ?? 0);
  const qU = U.length === 0 ? 0 : (U[0]?.length ?? 0);
  const qVt = Vt.length;
  const tolFactor = SAFETY * EPS * Math.max(m, n) * Math.sqrt(Math.min(m, n));

  const Sigma = buildSigma(S, qU, qVt);
  const reconstructed = matmul(matmul(U, Sigma), Vt);
  const aNorm = frobNorm(A);
  const err = frobNorm(matSub(reconstructed, A));

  if (aNorm === 0.0) {
    if (err > 0.0) {
      return { pass: false, detail: `||USVt − A||_F = ${fmt(err)} but A is zero` };
    }
    return { pass: true, detail: "zero input, zero residual" };
  }

  const rel = err / aNorm;
  if (rel > tolFactor) {
    return { pass: false, detail: `||USVt − A||_F / ||A||_F = ${fmt(rel)} > tol = ${fmt(tolFactor)}` };
  }
  return { pass: true, detail: `||USVt − A||_F / ||A||_F = ${fmt(rel)} ≤ tol = ${fmt(tolFactor)}` };
}

function checkSelfReportedResidual(
  A: number[][],
  U: number[][],
  S: number[],
  Vt: number[][],
  reported: number,
): Check {
  // Match verify.py: recomputed = ||USVt − A||_F / max(||A||_F, 1)
  const qU = U.length === 0 ? 0 : (U[0]?.length ?? 0);
  const qVt = Vt.length;
  const Sigma = buildSigma(S, qU, qVt);
  const reconstructed = matmul(matmul(U, Sigma), Vt);
  const aNorm = frobNorm(A);
  const err = frobNorm(matSub(reconstructed, A));
  const recomputed = err / Math.max(aNorm, 1.0);
  const diff = Math.abs(reported - recomputed);
  const rhs = SELF_REPORT_REL_TOL * Math.max(recomputed, EPS);
  if (diff > rhs) {
    return {
      pass: false,
      detail: `reported reconstruction_error=${reported.toExponential(6)} vs verifier=${recomputed.toExponential(6)}, |Δ|=${fmt(diff)} > tol=${fmt(rhs)}`,
    };
  }
  return {
    pass: true,
    detail: `reported=${reported.toExponential(6)}, verifier=${recomputed.toExponential(6)}`,
  };
}

function checkSelfReportedOrthogonality(
  U: number[][],
  Vt: number[][],
  reportedU: number,
  reportedVt: number,
): Check {
  const rU = frobNorm(gramMinusEye(U));
  const rV = frobNorm(rowGramMinusEye(Vt));
  const diffU = Math.abs(reportedU - rU);
  const diffV = Math.abs(reportedVt - rV);
  const rhsU = SELF_REPORT_REL_TOL * Math.max(rU, EPS);
  const rhsV = SELF_REPORT_REL_TOL * Math.max(rV, EPS);
  if (diffU > rhsU) {
    return {
      pass: false,
      detail: `orthogonality_error_U: reported=${reportedU.toExponential(6)} vs verifier=${rU.toExponential(6)}, |Δ|=${fmt(diffU)} > tol=${fmt(rhsU)}`,
    };
  }
  if (diffV > rhsV) {
    return {
      pass: false,
      detail: `orthogonality_error_Vt: reported=${reportedVt.toExponential(6)} vs verifier=${rV.toExponential(6)}, |Δ|=${fmt(diffV)} > tol=${fmt(rhsV)}`,
    };
  }
  return {
    pass: true,
    detail: `U: reported=${fmt(reportedU)}, verifier=${fmt(rU)}; Vt: reported=${fmt(reportedVt)}, verifier=${fmt(rV)}`,
  };
}

// ─── top-level verify ────────────────────────────────────────────────────────

function _wrap(checks: Record<string, Check>): { pass: boolean; reason: string; checks: Record<string, Check> } {
  const overall = Object.values(checks).every((c) => c.pass);
  if (overall) return { pass: true, reason: "all invariants hold", checks };
  const firstFail = Object.entries(checks).find(([, c]) => !c.pass)!;
  return { pass: false, reason: `failed: ${firstFail[0]} — ${firstFail[1].detail}`, checks };
}

function verify(payload: {
  input: { A: number[][]; mode?: string };
  candidate: unknown;
  id?: string;
}): { pass: boolean; reason: string; checks: Record<string, Check> } {
  const cand = payload.candidate;
  if (typeof cand !== "object" || cand === null) {
    return { pass: false, reason: "candidate must be a JSON object", checks: {} };
  }

  const mode = payload.input.mode ?? "reduced";
  if (mode !== "reduced" && mode !== "complete") {
    return { pass: false, reason: `unknown mode ${JSON.stringify(mode)}`, checks: {} };
  }

  const A = payload.input.A;
  if (!Array.isArray(A)) {
    return { pass: false, reason: "A must be a 2-D list", checks: {} };
  }

  const checks: Record<string, Check> = {};
  const candR = cand as Record<string, unknown>;

  // 1. shape (must come first; downstream checks dereference fields)
  checks["shape"] = checkShape(A, candR, mode);
  if (!checks["shape"].pass) return _wrap(checks);

  // 2. finite_entries (must come before any numerical checks)
  checks["finite_entries"] = checkFinite(candR);
  if (!checks["finite_entries"].pass) return _wrap(checks);

  const U  = candR["U"]  as number[][];
  const S  = candR["S"]  as number[];
  const Vt = candR["Vt"] as number[][];

  // 3. S_nonneg_descending
  checks["S_nonneg_descending"] = checkSNonnegDescending(S);

  // 4. U_orthonormal
  checks["U_orthonormal"] = checkUOrthonormal(A, U);

  // 5. Vt_orthonormal
  checks["Vt_orthonormal"] = checkVtOrthonormal(A, Vt);

  // 6. factorisation_residual
  checks["factorisation_residual"] = checkFactorisationResidual(A, U, S, Vt);

  // 7. self_reported_residual
  checks["self_reported_residual"] = checkSelfReportedResidual(A, U, S, Vt, candR["reconstruction_error"] as number);

  // 8. self_reported_orthogonality
  checks["self_reported_orthogonality"] = checkSelfReportedOrthogonality(
    U, Vt,
    candR["orthogonality_error_U"] as number,
    candR["orthogonality_error_Vt"] as number,
  );

  return _wrap(checks);
}

// ─── main ────────────────────────────────────────────────────────────────────

async function main(): Promise<void> {
  let stdin = "";
  for await (const chunk of Bun.stdin.stream()) {
    stdin += new TextDecoder().decode(chunk);
  }
  let result: { pass: boolean; reason: string; checks: Record<string, Check> };
  try {
    const payload = JSON.parse(stdin) as {
      input: { A: number[][]; mode?: string };
      candidate: unknown;
      id?: string;
    };
    result = verify(payload);
  } catch (e) {
    const err = e as Error;
    process.stderr.write(`${err.stack ?? err.message}\n`);
    result = { pass: false, reason: `verifier crashed: ${err.name}: ${err.message}`, checks: {} };
  }
  process.stdout.write(JSON.stringify(result) + "\n");
}

await main();

File: benchmarks/linalg-svd/golden/verify.py
#!/usr/bin/env python3
"""linalg-svd verifier — invariant-based, language-neutral.

stdin:
  {"input": {"A": [[float, ...], ...], "mode"?: "reduced" | "complete"},
   "candidate": {"U": ..., "S": ..., "Vt": ..., "mode": ...,
                 "reconstruction_error": ..., "orthogonality_error_U": ...,
                 "orthogonality_error_Vt": ..., "condition_number": ...,
                 "rank_estimate": ..., "method": ..., "warnings": [...]},
   "id"?: str}

stdout:
  {"pass": bool, "reason": str, "checks": {<name>: {"pass": bool, "detail": str}}}

Runs 8 independent checks per case (see verifier_protocol.md):
  1. shape                        — U, S, Vt dimensions match (m, n, mode)
  2. finite_entries               — no NaN, no ±Inf in factors or scalars
  3. S_nonneg_descending          — singular values ≥ 0 and non-increasing
  4. U_orthonormal                — ||UᵀU − I|| within tol_orth
  5. Vt_orthonormal               — ||Vt·Vtᵀ − I|| within tol_orth
  6. factorisation_residual       — ||U·diag(S)·Vt − A||_F within tol_recon
  7. self_reported_residual       — candidate's recon claim is honest
  8. self_reported_orthogonality  — candidate's orth claims (U + Vt) honest

Tolerances: Higham 2002 §20.3 / Demmel-Kahan 1990, with 100× empirical
safety factor (SciPy LAPACK clears each bound by ≥1 order of magnitude).
"""

from __future__ import annotations

import json
import sys
import traceback
from typing import Any

import numpy as np

EPS = 2.220446049250313e-16
SAFETY = 100.0
SELF_REPORT_REL_TOL = 1e-6


# --- helpers -----------------------------------------------------------------


def _is_finite_array(x: Any) -> bool:
    arr = np.asarray(x, dtype=float)
    return bool(np.all(np.isfinite(arr)))


def _shape_of(mat: Any) -> tuple[int, int] | None:
    if not isinstance(mat, list):
        return None
    if not mat:
        return (0, 0)
    if not all(isinstance(row, list) for row in mat):
        return None
    n_cols = len(mat[0])
    if not all(len(row) == n_cols for row in mat):
        return None
    if not all(isinstance(v, (int, float)) for row in mat for v in row):
        return None
    return (len(mat), n_cols)


def _vec_len(v: Any) -> int | None:
    if not isinstance(v, list):
        return None
    if not all(isinstance(x, (int, float)) for x in v):
        return None
    return len(v)


# --- per-check implementations ----------------------------------------------


def check_shape(A: np.ndarray, candidate: dict, mode: str) -> dict:
    m, n = A.shape
    k = min(m, n)
    expected_U = (m, k) if mode == "reduced" else (m, m)
    expected_Vt = (k, n) if mode == "reduced" else (n, n)
    expected_S_len = k

    required = {
        "U", "S", "Vt", "mode",
        "reconstruction_error", "orthogonality_error_U",
        "orthogonality_error_Vt", "condition_number",
        "rank_estimate", "method", "warnings",
    }
    missing = required - set(candidate.keys())
    if missing:
        return {"pass": False, "detail": f"missing fields: {sorted(missing)}"}

    if candidate["mode"] != mode:
        return {"pass": False,
                "detail": f"mode mismatch: candidate={candidate['mode']!r}, expected={mode!r}"}

    if not isinstance(candidate["method"], str):
        return {"pass": False, "detail": "method must be a string"}

    if not isinstance(candidate["warnings"], list) or \
       not all(isinstance(s, str) for s in candidate["warnings"]):
        return {"pass": False, "detail": "warnings must be list[str]"}

    if not isinstance(candidate["rank_estimate"], int) or \
       isinstance(candidate["rank_estimate"], bool) or \
       candidate["rank_estimate"] < 0:
        return {"pass": False, "detail": "rank_estimate must be a non-negative int"}

    sU = _shape_of(candidate["U"])
    if sU != expected_U:
        return {"pass": False,
                "detail": f"U shape {sU}, expected {expected_U} for mode={mode!r}"}

    sS = _vec_len(candidate["S"])
    if sS != expected_S_len:
        return {"pass": False,
                "detail": f"S length {sS}, expected {expected_S_len}"}

    sVt = _shape_of(candidate["Vt"])
    if sVt != expected_Vt:
        return {"pass": False,
                "detail": f"Vt shape {sVt}, expected {expected_Vt} for mode={mode!r}"}

    return {"pass": True, "detail": f"U {sU}, S {sS}, Vt {sVt}, mode {mode!r}"}


def check_finite(candidate: dict) -> dict:
    for field in ("U", "S", "Vt"):
        if not _is_finite_array(candidate[field]):
            return {"pass": False, "detail": f"non-finite entry in {field}"}
    for field in ("reconstruction_error", "orthogonality_error_U",
                  "orthogonality_error_Vt", "condition_number"):
        v = candidate[field]
        if not isinstance(v, (int, float)) or not np.isfinite(v):
            return {"pass": False, "detail": f"non-finite or non-numeric {field}"}
    return {"pass": True, "detail": "all entries finite"}


def check_S_nonneg_descending(S: np.ndarray) -> dict:
    if len(S) == 0:
        return {"pass": True, "detail": "empty S"}
    s_max = float(S[0]) if len(S) > 0 else 0.0
    tol_struct = SAFETY * EPS * max(s_max, 1.0)
    for i in range(len(S)):
        if S[i] < -tol_struct:
            return {"pass": False,
                    "detail": f"S[{i}]={S[i]:.3e} < -tol={-tol_struct:.3e} (negative singular value)"}
    for i in range(len(S) - 1):
        if S[i] + tol_struct < S[i + 1]:
            return {"pass": False,
                    "detail": (f"S[{i}]={S[i]:.3e} < S[{i+1}]={S[i+1]:.3e} "
                                f"(non-monotone, gap {S[i+1]-S[i]:.3e} > tol={tol_struct:.3e})")}
    return {"pass": True,
            "detail": f"S[0]={S[0]:.3e}, S[k-1]={S[-1]:.3e}, monotone within tol={tol_struct:.3e}"}


def check_U_orthonormal(A: np.ndarray, U: np.ndarray) -> dict:
    m, _ = A.shape
    q = U.shape[1]
    tol = SAFETY * EPS * m * np.sqrt(q)
    err = float(np.linalg.norm(U.T @ U - np.eye(q), ord="fro"))
    if err > tol:
        return {"pass": False,
                "detail": f"||UᵀU − I||_F = {err:.3e} > tol = {tol:.3e}"}
    return {"pass": True, "detail": f"||UᵀU − I||_F = {err:.3e} ≤ tol = {tol:.3e}"}


def check_Vt_orthonormal(A: np.ndarray, Vt: np.ndarray) -> dict:
    m, _ = A.shape
    q = Vt.shape[0]
    tol = SAFETY * EPS * m * np.sqrt(q)
    err = float(np.linalg.norm(Vt @ Vt.T - np.eye(q), ord="fro"))
    if err > tol:
        return {"pass": False,
                "detail": f"||Vt·Vtᵀ − I||_F = {err:.3e} > tol = {tol:.3e}"}
    return {"pass": True, "detail": f"||Vt·Vtᵀ − I||_F = {err:.3e} ≤ tol = {tol:.3e}"}


def check_factorisation_residual(A: np.ndarray, U: np.ndarray, S: np.ndarray,
                                  Vt: np.ndarray) -> dict:
    m, n = A.shape
    A_norm = float(np.linalg.norm(A, ord="fro"))
    tol_factor = SAFETY * EPS * max(m, n) * np.sqrt(min(m, n))

    Sigma = np.zeros((U.shape[1], Vt.shape[0]))
    for i in range(len(S)):
        Sigma[i, i] = S[i]
    err = float(np.linalg.norm(U @ Sigma @ Vt - A, ord="fro"))

    if A_norm == 0.0:
        if err > 0.0:
            return {"pass": False,
                    "detail": f"||USVt − A||_F = {err:.3e} but A is zero"}
        return {"pass": True, "detail": "zero input, zero residual"}

    rel = err / A_norm
    if rel > tol_factor:
        return {"pass": False,
                "detail": f"||USVt − A||_F / ||A||_F = {rel:.3e} > tol = {tol_factor:.3e}"}
    return {"pass": True,
            "detail": f"||USVt − A||_F / ||A||_F = {rel:.3e} ≤ tol = {tol_factor:.3e}"}


def check_self_reported_residual(A: np.ndarray, U: np.ndarray, S: np.ndarray,
                                  Vt: np.ndarray, reported: float) -> dict:
    A_norm = float(np.linalg.norm(A, ord="fro"))
    Sigma = np.zeros((U.shape[1], Vt.shape[0]))
    for i in range(len(S)):
        Sigma[i, i] = S[i]
    recomputed = float(np.linalg.norm(U @ Sigma @ Vt - A, ord="fro")) / max(A_norm, 1.0)
    diff = abs(reported - recomputed)
    rhs = SELF_REPORT_REL_TOL * max(recomputed, EPS)
    if diff > rhs:
        return {"pass": False,
                "detail": (f"reported reconstruction_error={reported:.6e} "
                           f"vs verifier={recomputed:.6e}, "
                           f"|Δ|={diff:.3e} > tol={rhs:.3e}")}
    return {"pass": True,
            "detail": f"reported={reported:.6e}, verifier={recomputed:.6e}"}


def check_self_reported_orthogonality(U: np.ndarray, Vt: np.ndarray,
                                       reported_U: float, reported_Vt: float) -> dict:
    qU = U.shape[1]
    qV = Vt.shape[0]
    rU = float(np.linalg.norm(U.T @ U - np.eye(qU), ord="fro"))
    rV = float(np.linalg.norm(Vt @ Vt.T - np.eye(qV), ord="fro"))

    diffU = abs(reported_U - rU)
    diffV = abs(reported_Vt - rV)
    rhsU = SELF_REPORT_REL_TOL * max(rU, EPS)
    rhsV = SELF_REPORT_REL_TOL * max(rV, EPS)
    if diffU > rhsU:
        return {"pass": False,
                "detail": (f"orthogonality_error_U: reported={reported_U:.6e} "
                           f"vs verifier={rU:.6e}, |Δ|={diffU:.3e} > tol={rhsU:.3e}")}
    if diffV > rhsV:
        return {"pass": False,
                "detail": (f"orthogonality_error_Vt: reported={reported_Vt:.6e} "
                           f"vs verifier={rV:.6e}, |Δ|={diffV:.3e} > tol={rhsV:.3e}")}
    return {"pass": True,
            "detail": f"U: reported={reported_U:.3e}, verifier={rU:.3e}; Vt: reported={reported_Vt:.3e}, verifier={rV:.3e}"}


# --- top-level verify --------------------------------------------------------


def verify(payload: dict) -> dict:
    inp = payload["input"]
    candidate = payload["candidate"]
    mode = inp.get("mode", "reduced")
    if mode not in ("reduced", "complete"):
        return {"pass": False, "reason": f"unknown mode {mode!r}", "checks": {}}

    if not isinstance(candidate, dict):
        return {"pass": False, "reason": "candidate must be a JSON object", "checks": {}}

    A = np.asarray(inp["A"], dtype=np.float64)
    if A.ndim != 2:
        return {"pass": False, "reason": f"A must be 2-D; got ndim={A.ndim}", "checks": {}}

    checks: dict[str, dict] = {}

    # 1. shape (must come first)
    checks["shape"] = check_shape(A, candidate, mode)
    if not checks["shape"]["pass"]:
        return _wrap(checks)

    # 2. finite_entries
    checks["finite_entries"] = check_finite(candidate)
    if not checks["finite_entries"]["pass"]:
        return _wrap(checks)

    U = np.asarray(candidate["U"], dtype=np.float64)
    S = np.asarray(candidate["S"], dtype=np.float64)
    Vt = np.asarray(candidate["Vt"], dtype=np.float64)

    # 3. S_nonneg_descending
    checks["S_nonneg_descending"] = check_S_nonneg_descending(S)

    # 4. U_orthonormal
    checks["U_orthonormal"] = check_U_orthonormal(A, U)

    # 5. Vt_orthonormal
    checks["Vt_orthonormal"] = check_Vt_orthonormal(A, Vt)

    # 6. factorisation_residual
    checks["factorisation_residual"] = check_factorisation_residual(A, U, S, Vt)

    # 7. self_reported_residual
    checks["self_reported_residual"] = check_self_reported_residual(
        A, U, S, Vt, float(candidate["reconstruction_error"]))

    # 8. self_reported_orthogonality
    checks["self_reported_orthogonality"] = check_self_reported_orthogonality(
        U, Vt, float(candidate["orthogonality_error_U"]),
        float(candidate["orthogonality_error_Vt"]))

    return _wrap(checks)


def _wrap(checks: dict) -> dict:
    overall = all(c["pass"] for c in checks.values())
    if overall:
        return {"pass": True, "reason": "all invariants hold", "checks": checks}
    first_fail = next(k for k, v in checks.items() if not v["pass"])
    return {"pass": False,
            "reason": f"failed: {first_fail} — {checks[first_fail]['detail']}",
            "checks": checks}


def main() -> None:
    try:
        payload = json.load(sys.stdin)
        result = verify(payload)
    except Exception as e:
        sys.stderr.write(traceback.format_exc())
        sys.stderr.write("\n")
        result = {
            "pass": False,
            "reason": f"verifier crashed: {type(e).__name__}: {e}",
            "checks": {},
        }
    json.dump(result, sys.stdout)
    sys.stdout.write("\n")


if __name__ == "__main__":
    main()
